### ISSUE TITLE
Missing value marker for 'onoff' columns

### ISSUE THREAD

**Issue by jfcorbett** [startables/startable-standard#4]
Missing value indicators (`'-'`, `'NaN'`...) are allowed in number columns and datetime columns. In 'text' columns, missing values are also allowed in the form of empty strings. 

Can we please also allow missing value indicators e.g. `'-'` in onoff (Boolean) columns? They are currently the odd one out with no provision for missing values.

**Comment by jfcorbett** [comment]
@JanusWesenberg ?

**Comment by JanusWesenberg** [comment]
This ties back to the question of column dtypes which we had somewhere else. My preference is that we should have dtype guarantees on columns: 
- dates should be datetime for which pandas has `pd.NaT` (don't know if `datetime` has something similar)
- text should be string for which empty and dash can be represented as-is
- onoff should be bool for which there is no default value
- everything else should be float for which there is nan. (This means casting some ints to float. If we want ints, we should introduce them).

So basically, my take would be for the reader to fail hard on missing bool values.

(BTW: should we make `bool` a synonym for `onoff`?)

**Comment by jfcorbett** [comment]
> should we make bool a synonym for onoff?

That would make sense to me. Now https://github.com/startable/startable-standard/issues/5

> onoff should be bool for which there is no default value

... certainly in Python and plain pandas. But, keeping in mind that StarTable is language-agnostic, is this universally the case? Apparently not.

[Here's an example](https://uwekorn.com/2019/09/02/boolean-array-with-missings.html) where pyarrow stores bools as 2 bits including a flag for present/missing; and is used to back a special implementation of bool Series in pandas. 

Also other languages support three-value logical data types and operations, [notably C#](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/bool#three-valued-boolean-logic). 

So to me the question is:
Is there a potential use case for null boolean values? I'm sure there is. But is there one that warrants:
* deviating from the perhaps more common, but evidently not universal, convention of there not being a "null value" for booleans, and
* going against the grain of Python/pandas to support it in the specific case immediately at hand i.e. pdtable development?

I'm on the fence. On the one hand I'd like to punt this one. On the other, implementing null bools now may save us some retrofitting later if we don't do it but later change our minds.

**Comment by JanusWesenberg** [comment]
Is this not flipping language agnosticism on its head? If nullable bool is only supported by some languages, should that not be an argument for using normal bool?

The practical reason I would like the `onoff` to be `bool` is that my prime use-case is to use the column value directly as a logical index. This is fine with a nullable bool type, but if the language does not support nullable bool (e.g. matlab), we would be forced to cast `NA` to `false`. This would cause information loss on read (with resulting unfaithful round-trip processing) and feature disparity between languages.

If we are willing to accept these drawbacks, a pragmatic solution could be something like
> If the system allows, `onoff` columns should map to a nullable bool and `-` should be parsed as `null`. If not `-` should be parsed as `false`.

FWIW, pandas allows `NA`-values in the `boolean` dtype [since v1.0][1].

[1]: https://pandas.pydata.org/docs/user_guide/boolean.html

**Comment by jfcorbett** [comment]
I guess it's a question of what we mean by language agnosticism. Do we mean: 
1. Must support lowest common denominator -- in which case we shouldn't even allow e.g. NaNs in float columns because FORTRAN 77 or some such; or, 
2. Here's the standard; how to support it in your favourite language/framework is your problem, but here are some implementations you may use

It's a spectrum, in which I lean more towards 2.

**Comment by JanusWesenberg** [comment]
Not quite true, nan is part of the [IEEE754] spec for floating point operations so there will always be nan for floats. This causes some pain in the rust community because it results in floats only being partially ordered.  See e.g. [this HN comment][1].

[IEEE754]: https://en.wikipedia.org/wiki/IEEE_754
[1]: https://news.ycombinator.com/item?id=9089112

**Comment by jfcorbett** [comment]
Wow, just read that HN thread, and the anguish is real! (pun intended) Not all the comments are rational, though. (that one too)

**Comment by jfcorbett** [comment]
Update: we have at least one active project that has gone ahead and used NaN markers `-` in `onoff` columns for some time. So the question is now: what hurts more? 
- Force that project to stop using NaN bools?
- Introduce NaN in `onoff`?

**Comment by JanusWesenberg** [comment]
Worrying about breaking bad implementations will not be a good way to define a useful standard. 

Implementing an efficient nullable bool is a major undertaking (see e.g. the [pandas docs](https://pandas.pydata.org/pandas-docs/stable/user_guide/boolean.html) or [this article](https://uwekorn.com/2019/09/02/boolean-array-with-missings.html)), and by including it in the standard, we force this work upon anyone wanting to implement a compliant reader.
A bit tempted to call feature-creep on this feature as well as the complex number one  #11 

If you really want nullable, you can make a column with a `.choice` template?

**Comment by jfcorbett** [comment]
So we're not convinced. Totally fair. 
In that case let's shelve this one.
